File: src/main/kotlin/ninja/blacknet/network/TxFetcher.kt
/*
 * Copyright (c) 2018-2019 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.network

import kotlinx.coroutines.channels.Channel
import ninja.blacknet.Runtime
import ninja.blacknet.core.TxPool
import ninja.blacknet.crypto.Hash
import ninja.blacknet.packet.GetTransactions
import ninja.blacknet.packet.Transactions
import ninja.blacknet.time.SystemClock
import ninja.blacknet.time.delay
import ninja.blacknet.time.milliseconds.MilliSeconds
import ninja.blacknet.util.SynchronizedHashMap

object TxFetcher {
    private val inventoryChannel: Channel<Pair<Connection, List<Hash>>> = Channel(Channel.UNLIMITED)
    private val requested = SynchronizedHashMap<Hash, MilliSeconds>()

    init {
        Runtime.rotate(::implementation)
        Runtime.rotate(::watchdog)
    }

    fun offer(connection: Connection, list: List<Hash>) {
        inventoryChannel.offer(Pair(connection, list))
    }

    suspend fun fetched(hash: Hash): Boolean {
        return requested.remove(hash) != null
    }

    private suspend fun implementation() {
        val (connection, inventory) = inventoryChannel.receive()

        val request = ArrayList<Hash>(inventory.size)
        val currTime = SystemClock.milliseconds

        for (hash in inventory) {
            if (requested.containsKey(hash)) {
                continue
            }

            if (TxPool.isInteresting(hash)) {
                requested.put(hash, currTime)
                request.add(hash)
            }

            if (request.size == Transactions.MAX) {
                sendRequest(connection, request)
                request.clear()
            }
        }

        if (request.size != 0) {
            sendRequest(connection, request)
        }
    }

    private fun sendRequest(connection: Connection, request: ArrayList<Hash>) {
        connection.sendPacket(GetTransactions(request))
    }

    private suspend fun watchdog() {
        delay(Node.NETWORK_TIMEOUT)

        val currTime = SystemClock.milliseconds
        val timeouted = requested.filterToKeyList { _, time -> currTime > time + Node.NETWORK_TIMEOUT }
        requested.removeAll(timeouted)
    }
}

File: network/src/main/rust/txfetcher.rs
/*
 * Copyright (c) 2018-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::connection::Connection;
use crate::node::NETWORK_TIMEOUT;
use crate::packet::{GetTransactions, Inventory, MAX_INVENTORY, MAX_TRANSACTIONS};
use crate::txpool::TxPool;
use blacknet_kernel::blake2b::Hash;
use blacknet_time::{Milliseconds, SystemClock};
use core::cmp::min;
use std::collections::HashMap;
use std::sync::{Arc, Mutex, RwLock, Weak};
use tokio::runtime::Runtime;
use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender, unbounded_channel};
use tokio::time::sleep;

const MAX_REQUESTS: usize = 16 * MAX_INVENTORY;

pub struct TxFetcher {
    inventory_send: UnboundedSender<(Weak<Connection>, Inventory)>,
    requests: Mutex<HashMap<Hash, (u64, Milliseconds)>>,
    tx_pool: Weak<RwLock<TxPool>>,
}

impl TxFetcher {
    pub fn new(runtime: &Runtime, tx_pool: Weak<RwLock<TxPool>>) -> Arc<Self> {
        let (inventory_send, inventory_recv) = unbounded_channel();
        let tx_fetcher = Arc::new(Self {
            inventory_send,
            requests: Mutex::new(HashMap::new()),
            tx_pool,
        });

        runtime.spawn(tx_fetcher.clone().implementation(inventory_recv));
        runtime.spawn(tx_fetcher.clone().watchdog());

        tx_fetcher
    }

    pub fn offer(&self, connection: Weak<Connection>, inventory: Inventory) {
        self.inventory_send.send((connection, inventory)).unwrap();
    }

    pub fn fetched(&self, connection: &Connection, hash: Hash) -> bool {
        let mut requests = self.requests.lock().unwrap();
        if let Some(&(id, _)) = requests.get(&hash)
            && connection.id() == id
        {
            return requests.remove(&hash).is_some();
        }
        false
    }

    async fn implementation(
        self: Arc<Self>,
        mut inventory_recv: UnboundedReceiver<(Weak<Connection>, Inventory)>,
    ) {
        loop {
            let (connection, inventory) = inventory_recv.recv().await.unwrap();
            let connection = if let Some(connection) = connection.upgrade() {
                connection
            } else {
                continue;
            };

            let mut requests = self.requests.lock().unwrap();
            if requests.len() >= MAX_REQUESTS {
                continue;
            }

            let capacity = min(inventory.len(), MAX_TRANSACTIONS);
            let mut request = GetTransactions::with_capacity(capacity);
            let tx_pool_strong = self.tx_pool.upgrade().unwrap();
            let tx_pool = tx_pool_strong.read().unwrap();
            let now = SystemClock::millis();

            for hash in inventory {
                if requests.contains_key(&hash) {
                    continue;
                }

                if tx_pool.is_interesting(hash) {
                    requests.insert(hash, (connection.id(), now));
                    request.push(hash);
                }

                if request.len() == MAX_TRANSACTIONS {
                    connection.send_packet(&request);
                    request.clear();
                }
            }

            if !request.is_empty() {
                connection.send_packet(&request);
            }
        }
    }

    async fn watchdog(self: Arc<Self>) {
        loop {
            sleep(NETWORK_TIMEOUT.try_into().unwrap()).await;

            let mut requests = self.requests.lock().unwrap();
            let now = SystemClock::millis();
            requests.retain(|_, &mut (_, time)| now < time + NETWORK_TIMEOUT);
        }
    }
}
